### PR TITLE
Implement Decimal Types as Logical Types

### PR DESCRIPTION
**Pull request by majetideepak**
The scope of this PR is to address a couple of correlated enhancements for Decimal Types listed below
1) Introduced `TypeKind::HUGEINT` (`HugeintType`) with `int128_t` as its CPP type.
 The support is limited to the needs of the LongDecimalType.
2) Removed the `TypeKind::SHORT_DECIMAL`, `TypeKind::LONG_DECIMAL` enum values and 
replaced them with a Velox type ShortDecimalType based on BigintType and LongDecimalType 
based on HugeintType. This change replaces Decimal CPP types `UnscaledShortDecimal` with `int64_t`, 
`UnscaledLongDecimal` with `int128_t`.
3) Removed `SHORT_DECIMAL`, `LONG_DECIMAL` APIs and replaced them with `DECIMAL`.

The above changes mean data of decimal types are stored in memory using 64-bit/128-bit integers.
 e.g: `FlatVector<int64_t>`, `FlatVector<int128_t>`. The individual value is an unscaled decimal value. 
A variant similarly holds an unscaled value using 64-bit/128-bit integers.
The Decimal Type must be present to interpret the decimal semantics.

Earlier, the decimal limit check was enabled in UnscaledXXXDecimal only in debug mode. 
That won't catch overflows in production (release build). The new approach would be to explicitly call `DecimalUtil::valueInRange(int128_t)`where ever a decimal value is computed. 
See usage in SumAggregate.h, DecimalArithmetic.cpp. 
This is what the Presto Java implementation does as well.

Vector functions must use `DECIMAL` in the signature. 

Simple functions require a unique type for ShortDecimal and LongDecimal to bind the appropriate
implementation. Therefore, simple functions are not supported for Decimal Types.

See https://github.com/facebookincubator/velox/discussions/4069

### CODE REVIEW

**Comment by majetideepak** [comment]
@mbasmanova I still have some refactoring. A big TODO is supporting Decimal variants on top of `BIGINT` and `HUGEINT`.
Have you thought of how logical types would be extended to variants? My initial thought is to serialize the type information as part of the variant `payload`.

**Comment by mbasmanova** [comment]
@majetideepak 

> Have you thought of how logical types would be extended to variants? 

Deepak, in my proposal variant represents only physical value and requires a logic type to interpret it. Therefore, there will be no decimal variant, but only 64-bit and 128-bit integer variants.

File: velox/type/HugeInt.h [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+#include "velox/common/base/BitUtil.h"
+#include "velox/common/base/Exceptions.h"
+#include "velox/type/StringView.h"
+
+#pragma once
+
+namespace facebook::velox {
+
+using int128_t = __int128_t;
+class HugeInt {
```
[the remark this answers is not in the document]
Comment (majetideepak) [reply velox/type/HugeInt.h:28]: Earlier, the limit check was enabled in UnscaledXXXDecimal only in debug mode. That won't catch overflows in production (release build). The new approach would be to explicitly call `DecimalUtil::valueInRange(int128_t)` where ever a decimal value is computed. See usage in `SumAggregate.h`. This is what the Presto Java implementation does as well.

**Comment by majetideepak** [comment]
@karteekmurthys The HUGEINT implementation is specific to LongDecimals. A generic HUGEINT support needs more work and would be a separate PR. I will try to split this PR into smaller pieces where possible.

**Review by karteekmurthys — APPROVED** [review]

File: velox/type/Type.h [hunk 2]
```diff
@@ -658,34 +634,75 @@ class DecimalType : public ScalarType<KIND> {
...
+  return dynamic_cast<const ShortDecimalType*>(&type) != nullptr;
 }
 
-bool isDecimalName(const std::string& typeName);
+FOLLY_ALWAYS_INLINE bool isLongDecimalType(const Type& type) {
+  return dynamic_cast<const LongDecimalType*>(&type) != nullptr;
+}
 
-bool isDecimalTypeSignature(const std::string& arg);
+FOLLY_ALWAYS_INLINE bool isDecimalType(const Type& type) {
```
Comment (karteekmurthys) [inline velox/type/Type.h:699]: Do you think checking (kindName() == DECIMAL && TypeKind == HUGEINT) for isDecimalType would be cleaner versus checking for dynamic casts?

@mbasmanova Are we going to have PhysicalTypeKind and LogicalTypeKind. For instance, LONG_DECIMAL is Logical type and the underlying physical type is HUGEINT.

File: velox/type/Type.h [hunk 3]
```diff
@@ -658,34 +634,75 @@ class DecimalType : public ScalarType<KIND> {
     return scale_;
   }
 
+  const char* kindName() const override {
+    return "DECIMAL";
```
Comment (mbasmanova) [inline velox/type/Type.h:638]: Shouldn't this be `return TypeTraits<KIND>::name;`, e.g. HUGEINT or BIGINT?

File: velox/type/Type.h (hunk 2 above)
Comment (mbasmanova) [reply velox/type/Type.h:699]: @karteekmurthys I'm working through converting INTERVAL_DAY_TIME to logical type in #4406.

My current thinking is that TypeKind enum describes physical type and an instance of Type class describes logical type. I do not plan to have enums for logical types as these should be extensible.

I expect LongDecimalType to be logical type backed by TypeKind::HUGEINT physical type and ShortDecimalType to be logical type backed by TypeKind::BIGINT physical type.

File: velox/type/Type.h (hunk 3 above)
Comment (majetideepak) [reply velox/type/Type.h:638]: I think the function name should be changed from`kindName` -> `name`
I need this to satisfy the following check-in the SignatureBinder. Without this, a decimal function can be bound to `bigint` arguments.
```
if (typeName != actualType->kindName()) {
    return false;
  }
```

Comment (majetideepak) [reply velox/type/Type.h:638]: Or we will need another API to get the logical name.

Comment (mbasmanova) [reply velox/type/Type.h:638]: In #4406, I introduced Type::name() for logical type name. Type::kindName() would continue to represent the physical type name.

Comment (mbasmanova) [reply velox/type/Type.h:638]: @majetideepak I extracted these changes into a separate PR https://github.com/facebookincubator/velox/pull/4519

**Comment by majetideepak** [comment]
@mbasmanova I don't think there is an alternative to support simple functions without the CustomTypes. The alternative is to use Vector Functions. 
We should make a decision soon since there is a lot more decimal type activity.

**Comment by mbasmanova** [comment]
> @mbasmanova I don't think there is an alternative to support simple functions without the CustomTypes. The alternative is to use Vector Functions. We should make a decision soon since I see a lot more decimal type activity.

@majetideepak Deepak, good morning. I discussed this with @laithsakka last night and out conclusion is that it would be too much work to support SimpleFunctions that use logical types with multiple physical backings. Hence, we suggest to re-write existing simple functions that use decimals into vector functions and write new functions as vector functions.

https://github.com/facebookincubator/velox/discussions/4069#discussioncomment-5615969

What do you think?

**Comment by majetideepak** [comment]
> too much work to support SimpleFunctions that use logical types with multiple physical backings. Hence, we suggest to re-write existing simple functions that use decimals into vector functions and write new functions as vector functions

I agree with this. Do you suggest we write the existing decimal simple functions as Vector Functions first?

**Review by mbasmanova — COMMENTED** [review]
@Yuhta Jimmy, would you help review this PR.

@majetideepak Deepak, this PR appears quite large. Is there any chance to break it down into smaller pieces?

**Comment by majetideepak** [comment]
@mbasmanova Most of the changes are simply replacing UnscaledXXXDecimal with int64_t/int128_t and removing TypeKinds SHORT_DECIMAL and LONG_DECIMAL. All the changes are related to the new logical type and removing the existing Decimal TypeKinds. I want to avoid having both implementations at the same time.
@Yuhta I am happy to help with the review process in any way I can. Please let me know.

File: velox/dwio/common/ScanSpec.cpp [hunk 4]
```diff
@@ -337,6 +337,9 @@ bool testFilter(
   if (mayHaveNull && filter->testNull()) {
     return true;
   }
+  if (type->isDecimal()) {
+    return true;
```
[the remark this answers is not in the document]
Comment (majetideepak) [reply velox/dwio/common/ScanSpec.cpp:341]: The stats from Parquet are in FixedLengthByteArray big-endian format. We need to convert them to BIGINT/HUGEINT.
I opened a PR https://github.com/facebookincubator/velox/issues/4779
This is consistent with the existing behavior where TypeKind::XXX_DECIMAL are not supported here.

File: velox/core/SimpleFunctionMetadata.h [hunk 5]
```diff
@@ -429,6 +429,11 @@ class SimpleFunctionMetadata : public ISimpleFunctionMetadata {
         std::move(results.stats)};
   }
 
+  FOLLY_ALWAYS_INLINE bool isDecimalTypeSignature(const std::string& arg) {
```
[the remark this answers is not in the document]
Comment (majetideepak) [reply velox/core/SimpleFunctionMetadata.h]: Removed.

File: velox/type/HugeInt.h [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+  build(uint64_t hi, uint64_t lo) {
+    // GCC does not allow left shift negative value.
+    return (static_cast<__uint128_t>(hi) << 64) | lo;
+  }
+
+  static constexpr FOLLY_ALWAYS_INLINE uint64_t lower(int128_t value) {
+    return static_cast<uint64_t>(value);
+  }
+
+  static constexpr FOLLY_ALWAYS_INLINE int64_t upper(int128_t value) {
```
[the remark this answers is not in the document]
Comment (majetideepak) [reply velox/type/HugeInt.h]: fixed.

File: velox/expression/CastExpr.h [hunk 7]
```diff
@@ -171,7 +171,7 @@ class CastExpr : public SpecialForm {
   /// @param rows Non-null rows of the input vector.
   /// @param input The input decimal vector. It is guaranteed to be flat or
   /// constant.
-  template <typename DecimalType>
+  template <typename toDecimalType>
```
[the remark this answers is not in the document]
Comment (majetideepak) [reply velox/expression/CastExpr.h]: fixed

File: velox/type/Variant.cpp [hunk 8]
```diff
@@ -223,7 +187,7 @@ void variant::throwCheckPtrError() const {
   throw std::invalid_argument{"missing variant value"};
 }
 
-std::string variant::toJson() const {
+std::string variant::toJson(TypePtr type) const {
```
[the remark this answers is not in the document]
Comment (majetideepak) [reply velox/type/Variant.cpp]: fixed

File: velox/type/HugeInt.h [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+      char* serializedData) {
+    memcpy(serializedData, &value, sizeof(int128_t));
+  }
+};
+
+} // namespace facebook::velox
+
+namespace folly {
+template <>
+struct hasher<__int128_t> {
```
[the remark this answers is not in the document]
Comment (majetideepak) [reply velox/type/HugeInt.h:61]: In Velox OSS, for some reason, the Folly dependency is not built with `FOLLY_HAVE_INT128_T` defined. That is why this redefinition is needed here. CI passes here without a clash. Is it different inside Meta? Should we enable it in OSS as well?

[the remark this answers is not in the document]
Comment (Yuhta) [reply velox/type/HugeInt.h:61]: Yes we should enable it

File: velox/type/CMakeLists.txt [hunk 10]
```diff
@@ -16,6 +16,7 @@ if(${VELOX_BUILD_TESTING})
 endif()
 add_subdirectory(tz)
 
+add_compile_definitions(FOLLY_HAVE_INT128_T=1)
```
Comment (majetideepak) [inline velox/type/CMakeLists.txt:19]: @Yuhta Modifying the Folly build seems to be non-trivial. Can we work with this approach?

Comment (Yuhta) [reply velox/type/CMakeLists.txt:19]: Yeah should be ok.  Can we move this to project root though?

Comment (majetideepak) [reply velox/type/CMakeLists.txt]: Done. Moving to the project root was required actually.

**Comment by Yuhta** [comment]
@majetideepak Can you double check if the failure in unit test is related?  There is a crash in `HashJoinTest.dynamicFiltersWithSkippedSplits`: https://app.circleci.com/pipelines/github/facebookincubator/velox/24145/workflows/49161f0c-82be-4ac9-91c4-857454ac6d52/jobs/153309

**Comment by majetideepak** [comment]
@Yuhta the debug build (linux-build CI) tests passed. I am re-running the tests again.

**Pull request closed without merging**